### ISSUE TITLE
Unassociated buildpacks cannot be updated to use a different stack without re-uploading them

### ISSUE THREAD

**Issue by sclevine** [cloudfoundry/cli#1493]
Cloud Controller allows system buildpacks that are not associated with a stack to be associated with a stack. 

This is documented here: https://apidocs.cloudfoundry.org/5.5.0/buildpacks/assign_buildpack_to_stack.html

This functionality is not currently exposed in the cf CLI.

**What's the user value of this feature request?**
Example use cases:

1. An operator has uploaded a custom buildpack intended to be used with the cflinuxfs2 stack. The operator upgrades their platform to a version of Cloud Controller that supports buildpack-to-stack association and adds the cflinuxfs3 stack to their foundation. The custom buildpack is not associated with any stack. The operator is unable to associate that buildpack with cflinuxfs2 without possessing a local copy of the buildpack (and it is not currently possible to download buildpacks from CFAR once they are uploaded). This means that the operator is unable to provide a cflinuxfs3 version of the buildpack, and any apps using the unassociated buildpack will fail to restage or launch if they are switched to the cflinuxfs3 stack. Additionally, this cflinuxfs2 buildpack will be included in the auto-detection process for cflinuxfs3 as well as Windows stacks.

2. An operator wishes to upload a particular version of the Java buildpack with a new name and associate it with cflinuxfs3-only. The Java buildpack is only offered unassociated, so they can only upload a copy that's unassociated. This operator is unable to prevent developers from staging the app with cflinuxfs2, and the auto-detection process for cflinuxfs2 as well as Windows stacks will include this buildpack.

**Who is the functionality for?**
Operators who manage buildpacks on a deployment of Cloud Foundry.

**How often will this functionality be used by the user?**
Whenever they manage buildpacks.

**Who else is affected by the change?**
This is a new feature that would not introduce any breaking changes.

**Is your feature request related to a problem? Please describe.**
See user value above.

**Describe the solution you'd like**

**When** The command
```
cf update-buildpack my_buildpack -s cflinuxfs3
```
is applied to an unassociated buildpack called `my_buildpack`
**And** no replacement buildpack is specified with `-p`
**And** another buildpack called `my_buildpack` is not already associated with cflinuxfs3
**Then** `my_buildpack` is associated with cflinuxfs3

**Describe alternatives you've considered**
To avoid overloading the `-s` flag, a new command could be introduced that just updates the stack:
```
cf update-buildpack-stack my_unassociated_buildpack cflinuxfs3
```

I have no preference between these two options. I have no UX preferences other than that the cf CLI allows an unassociated buildpack to be associated with a stack without re-uploading it.

CC: @dieucao @nebhale @abbyachau

**Comment by abbyachau** [comment]
Hi @sclevine could you pls confirm the following use cases:

1. Scenario: user is on 2.2, upgrades to 2.3, and they have their buildpacks bits accessible locally AND they have a buildpack manifest associated with a stack
---> current solution:  some of their buildpacks have a `nil` stack, they run `cf update-buildpack buildpack-with-nil-stack -s cflinuxfs3`

2. Scenario: user is on 2.2, upgrades to 2.3; and user no longer have their buildpack bits
---> current solution:  some of their buildpacks have a `nil` stack, they have to use the api directly to get out of this transitional state

3. Scenario: user is on 2.2, upgrades to 2.3; user has their buildpack bits saved locally BUT they are using a buildpack that has a unassociated stack
---> current solution:  some of their buildpacks have a `nil` stack, they have to use the api directly to get out of this transitional state

Does this describe the use cases? Are we missing any?

Also could you tell us more about the "unassociated stack" case. Also is this common? ( #1492)

**Comment by sclevine** [comment]
For (1), they would need to specify the buildpack bits in addition to the stack with the `-p` flag.

For (2), 👍 

For (3), 👍 

> Also could you tell us more about the "unassociated stack" case. 

In the unassociated state, buildpacks will work with any stack. However, this can break detection for all stacks where the unassociated buildpacks would pass detection when used with the wrong stack. For example, if an unassociated my_ruby_buildpack for cflinuxfs2 sits above an associated ruby_buildpack for cflinuxfs3, `cf push` would not work any ruby apps with cflinuxfs3, unless the app developer manually specified the buildpack.

The unassociated state also has other edge cases (e.g., it blocks uploading any new buildpacks with the same name). I think the CAPI team wants to deprecate the unassociated state completely in the long run.

> How were buildpacks with unassociated stacks working before, i.e. was defaulting to cflinuxfs2?

Unassociated buildpacks generally didn't cause issues before, because cflinuxfs3 wasn't available.

> Do you recall why we didn't default all nil stacks to cflinuxfs2?

Some buildpacks are intended to be used with windows stacks. Some CF vendors ship non-cflinuxfsX linux stacks (e.g., SUSE). We did consider this option though.

> Should/would it also help if there be a api endpoint to download their buildpacks?

Given that there's already an API endpoint to associate buildpacks, I don't think this would help in this particular case. But it is a request that I've heard a few times.

**Comment by abbyachau** [comment]
Thanks, I've added some acceptance criteria to the tracker story and will be scoping the flag in additional stories.

**Comment by abbyachau** [comment]
Hi @sclevine thanks for creating this issue. We still believe that there should be an endpoint to support downloading one's buildpacks should app operators find themselves losing and requiring their buildpack bits, (cc @tcdowney @charleshansen).

We've released this feature in [CLI v6.41](https://github.com/cloudfoundry/cli/releases/tag/v6.41.0), please do not hesitate to get in touch if you have any feedback. Thanks.
